basalt/src/config/keybinding.rs

```diff
@@ -0,0 +1,182 @@
+use std::{error::Error, fmt};
+
+use crossterm::event::{KeyCode, KeyEvent, KeyModifiers};
+use serde::{
+    de::{self, Visitor},
+    Deserialize, Deserializer,
+};
+
+use crate::app::{Action, ScrollAmount};
+
+#[derive(Clone, Debug, PartialEq, Deserialize)]
+pub(crate) struct KeyBinding {
+    pub key: Key,
+    pub command: Command,
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct Key {
+    pub modifiers: KeyModifiers,
+    pub code: KeyCode,
+}
+
+impl Key {
+    pub const CTRLC: Key = Key {
+        modifiers: KeyModifiers::CONTROL,
+        code: KeyCode::Char('c'),
+    };
+}
+
+impl<'de> Deserialize<'de> for Key {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_str(KeyVisitor)
+    }
+}
+
+struct KeyVisitor;
+
+impl Visitor<'_> for KeyVisitor {
+    type Value = Key;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a string whose format is either 'key' or 'modifier+key'")
+    }
+
+    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        let value = value.to_lowercase();
+        let (modifiers, code) = value.split_once('+').unwrap_or(("", value.as_str()));
+
+        Ok(Key {
+            modifiers: parse_modifiers(modifiers).map_err(de::Error::custom)?,
+            code: parse_code(code).map_err(de::Error::custom)?,
+        })
+    }
```

Comment:
I think a quite idiomatic pattern here would be to implement the `TryFrom` trait for the Key struct!

```suggestion
        value.try_into().map_err(de::Error::custom)
```

We hide the complexity in the try_from/try_into and can do the conversion directly on the `&str`.

```rust
impl TryFrom<&str> for Key {
    type Error = ConfigError;

    fn try_from(value: &str) -> Result<Self, Self::Error> {}
}
```

Reply:
Can you elaborate more on this point? I don't fully understand what you mean. Did you mean you tried implementing try_from, and it didn't work out?

Reply:
It did work, pretty much as the current implementation of Deserialize does.
But it felt cumbersome to Deserialize as a String to then .parse() it into a Key, so I invoked the function directly with #[serde(deserialize_with = "...")] to eventually write a Visitor, since I felt like the FromStr implementation would be used only to parse the configuration, and nowhere else.
At that point, I thought about more complex key bindings, and created the parse_modifiers and parse_code function, so the current model could be extended if need be later on.
It felt like the most streamlined way to go about it.